libmamba/include/mamba/core/satisfiability_error.hpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2019, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_PROBLEMS_GRAPH_HPP
+#define MAMBA_PROBLEMS_GRAPH_HPP
+
+#include <string>
+#include <utility>
+#include <variant>
+#include <unordered_map>
+
+#include "mamba/core/graph_util.hpp"
+#include "mamba/core/package_info.hpp"
+
+namespace mamba
+{
+
+    class MSolver;
+    class MPool;
+
+    /**
+     * Separate a dependency spec into a package name and the version range.
+     */
+    class DependencyInfo
+    {
+    public:
+        DependencyInfo(const std::string& dependency);
+
+        const std::string& name() const;
+        const std::string& range() const;
+        std::string str() const;
+
+    private:
+        std::string m_name;
+        std::string m_range;
+    };
+
+    /**
+     * A directed graph of the pacakges involved in a libsolv conflict.
+     */
+    class ProblemsGraph
+    {
+    public:
+        /**
+         * A simplification of the libsolv SolverRuleinfo
+         */
+        enum class ProblemType
+        {
+            CONFLICT,
+            NOT_FOUND,
+            NOT_INSTALLABLE,
+            BEST_NOT_INSTALLABLE,
+            ONLY_DIRECT_INSTALL,
+            EXCLUDED_BY_REPO_PRIORITY,
+            INFERIOR_ARCH,
+            PROVIDED_BY_SYSTEM
+        };
+
+        struct PackageNode
+        {
+            PackageInfo package_info;
+            std::optional<ProblemType> problem_type;
+        };
+        struct UnresolvedDependencyNode
+        {
+            std::string dependency;
+            std::optional<ProblemType> problem_type;
+        };
+        struct RootNode
+        {
+        };
+        using node_t = std::variant<PackageNode, UnresolvedDependencyNode, RootNode>;
+
+        struct RequireEdge : DependencyInfo
+        {
+        };
+        struct ConstraintEdge : DependencyInfo
+        {
+        };
+        using edge_t = std::variant<RequireEdge, ConstraintEdge>;
+
+        using graph_t = DiGraph<node_t, edge_t>;
```

Comment:
after the creation of the initial problems graph we would need to merge the graph and I think we should have a generic ProblemsGraph that will be the result of the merging as well. 
In the previous implementation I had different node & edge types for the initial graph and the merged graph. If we do that, we would need to have these classes outside of the ProblemsGraph and the class templated. What do you think ?

Reply:
I was thinking the same. This class is very small though (mostly aliases), it won't be difficult to extend.

My opinion is we should finish testing it and get it reviewed/merged. We can make the the changes in another PR once we defined the Node/Edge types for the compressed graph.

Reply:
Sounds good!